R/zchunk_LB134.Diet_Rfao.R

```diff
@@ -0,0 +1,380 @@
+                             value)) %>%
+      # interpolate out to that convergence year
+      group_by(GCAM_region_ID, GCAM_demand) %>%
+      mutate(value = approx_fun(year, value)) %>%
+      filter(year %in% c(HISTORICAL_YEARS, FUTURE_YEARS)) ->
+      L134.pcFood_kcald_R_Dmnd_Y
+
+    # Alternative diet scenarios for the SSPs
+    # Original lines 140-
+    # this same logic happens over and over, so *use a function*
+    # scen - scenario
+    create_ssp <- function(L102.pcgdp_thous90USD_Scen_R_Y, scen, demand, L134.pcFood_kcald_R_Dmnd_Y, A_FoodDemand_SSPs) {
+      if(demand == "crops") {
+        a <- 4545
+        b <- -0.099
+      } else if(demand == "meat") {
+        a <- 818
+        b <- -2.31
+      } else {
+        stop("Unknown demand ", demand)
+      }
+      L102.pcgdp_thous90USD_Scen_R_Y %>%
+        filter(scenario == scen) %>%
+        mutate(GCAM_demand = demand,
+               value = a / (1 + exp(b * log(value)))) %>%
+        select(GCAM_region_ID, GCAM_demand, year, value) %>%
+        filter(year %in% c(max(HISTORICAL_YEARS), FUTURE_YEARS)) ->
+        ssp_gdp
+
+      ssp_gdp %>%
+        # compute gdp ratios for future years
+        arrange(GCAM_region_ID, GCAM_demand, desc(year)) %>%
+        group_by(GCAM_region_ID, GCAM_demand) %>%
+        mutate(ratio = value / lead(value)) %>%
+        select(-value) %>%
+        # ...and use those ratios to scale future food demand
+        left_join(filter(L134.pcFood_kcald_R_Dmnd_Y, GCAM_demand == demand), ., by = c("GCAM_region_ID", "GCAM_demand", "year")) %>%
+        mutate(ratio = if_else(is.na(ratio), 1.0, ratio)) %>%
+        arrange(GCAM_region_ID, GCAM_demand, year) %>%
+        group_by(GCAM_region_ID, GCAM_demand) %>%
+        # computing the cumulative product, and then multiplying by the last historical year value, is the
+        # same mathematically as multiplying each year's ratio by the previous year's value
+        mutate(ratio = cumprod(ratio),
+               value = if_else(year > max(HISTORICAL_YEARS), ratio * nth(value, which(year == max(HISTORICAL_YEARS))), value)) %>%
+        # ensure the year-to-year change, and absolute values, don't exceed certain levels (given in 'A_FoodDemand_SSPs')
+        mutate(scenario = scen) %>%
+        left_join_error_no_match(A_FoodDemand_SSPs, by = c("scenario", "GCAM_demand")) ->
+        ssp_gdp
+
+      # Cap future values, both maximum annual change (ratio)
+      # and absolute value, based on values in A_FoodDemand_SSPs
+      # Because of the dependencies involved, I couldn't figure out a way
+      # not to use a loop here.
+
+      # Note that the old code has a bug: in line 168 a loop starts without
+      # correctly resetting the `prev_i` variable; as a result, if the value in
+      # the final historical year is 0, incorrect values get written to all future
+      # years -- but only for meat. We replicate this craptastic behavior below.
```

Comment:
I literally do not care and you are the ultimate arbiter of what goes in the DSR, but maybe "faulty behavior"?